macaroonbakery/httpbakery/error.py

```diff
@@ -54,14 +79,100 @@ def request_version(req_headers):
         x = int(vs)
     except ValueError:
         # Badly formed header - use backward compatibility mode.
-        return macaroonbakery.BAKERY_V1
-    if x > macaroonbakery.LATEST_BAKERY_VERSION:
+        return BAKERY_V1
+    if x > LATEST_BAKERY_VERSION:
         # Later version than we know about - use the
         # latest version that we can.
-        return macaroonbakery.LATEST_BAKERY_VERSION
+        return LATEST_BAKERY_VERSION
     return x
+
+
+class Error(namedtuple('Error', 'code, message, version, info')):
+    @classmethod
+    def deserialize(cls, serialized):
+        code = serialized.get('Code')
+        message = serialized.get('Message')
+        info = ErrorInfo.deserialize(serialized.get('Info'))
+        return Error(code=code, message=message, info=info,
+                     version=LATEST_BAKERY_VERSION)
+
+    def interaction_method(self, kind, x):
+        ''' Checks whether the error is an InteractionRequired error
+        that implements the method with the given name, and JSON-unmarshals the
+        method-specific data into x.
+        '''
+        if self.info is None or self.code != ERR_INTERACTION_REQUIRED:
+            raise InteractionError(
+                'not an interaction-required error (code {})'.format(
+                    self.code)
+            )
+        entry = self.info.interaction_methods.get(kind)
+        if entry is None:
+            raise InteractionMethodNotFound(
+                'interaction method {} not found'.format(kind)
+            )
+        return x.deserialize(entry)
+
+
+class ErrorInfo(
+    namedtuple('ErrorInfo', 'macaroon, macaroon_path, cookie_name_suffix, '
+                            'interaction_methods, visit_url, wait_url')):
+    '''  Holds additional information provided
+    by an error.
+
+    @param macaroon may hold a macaroon that, when
+    discharged, may allow access to a service.
+    This field is associated with the ERR_DISCHARGE_REQUIRED
+    error code.
+
+    @ param macaroon_path holds the URL path to be associated
+    with the macaroon. The macaroon is potentially
+    valid for all URLs under the given path.
+    If it is empty, the macaroon will be associated with
+    the original URL from which the error was returned.
+
+    @param cookie_name_suffix holds the desired cookie name suffix to be
+    associated with the macaroon. The actual name used will be
+    ("macaroon-" + cookie_name_suffix). Clients may ignore this field -
+    older clients will always use ("macaroon-" +
+    macaroon.signature() in hex).
+
+    @param visit_url holds a URL that the client should visit
+    in a web browser to authenticate themselves.
+
+    @param wait_url holds a URL that the client should visit
+    to acquire the discharge macaroon. A GET on
+    this URL will block until the client has authenticated,
+    and then it will return the discharge macaroon.
+    '''
+
+    __slots__ = ()
+
+    @classmethod
+    def deserialize(cls, serialized):
```

Comment:
Again, please document this method, and I'm not sure that deserialize is the right name here.

Reply:
done